components/scream/src/physics/shoc/shoc_linear_interp_impl.hpp

```diff
@@ -13,10 +13,50 @@ namespace shoc {
 
 template<typename S, typename D>
 KOKKOS_FUNCTION
-void Functions<S,D>::linear_interp(const uview_1d<const Spack>& x1, const uview_1d<const Spack>& x2, const uview_1d<const Spack>& y1, const uview_1d<Spack>& y2, const Int& km1, const Int& km2, const Int& ncol, const Spack& minthresh)
+void Functions<S,D>::linear_interp(
+  const MemberType& team,
+  const uview_1d<const Spack>& x1,
+  const uview_1d<const Spack>& x2,
+  const uview_1d<const Spack>& y1,
+  const uview_1d<Spack>& y2,
+  const Int& km1,
+  const Int& km2,
+  const Scalar& minthresh)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
+  const auto sx1 = scalarize(x1);
+  const auto sy1 = scalarize(y1);
+  const Int km2_pack = ekat::npack<Spack>(km2);
+
+  if (km1 == km2+1) {
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(team, km2_pack), [&] (const Int& k2) {
+      Spack x1, x1s, y1, y1s; // s->-1 shift
+      auto indx_pack = ekat::range<IntSmallPack>(k2*Spack::n + 1);
+      ekat::index_and_shift<-1>(sx1, indx_pack, x1, x1s);
+      ekat::index_and_shift<-1>(sy1, indx_pack, y1, y1s);
+      y2(k2) = y1s + (y1-y1s)*(x2(k2)-x1s)/(x1-x1s);
+    });
+  }
+  else if (km2 == km1+1) {
+
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(team, km2_pack), [&] (const Int& k2) {
+      Spack x1, x1s, y1, y1s; // s->-1 shift
+      auto indx_pack = ekat::range<IntSmallPack>(k2*Spack::n);
+      indx_pack.set(indx_pack < 1, 1); // special shift for 0 boundary case
```

Comment:
I recommend you separate out the top and bottom from the interior to avoid this special casing in the interior.